libraries/standard/mqtt/src/mqtt.c

```diff
@@ -375,41 +438,41 @@ MQTTStatus_t MQTT_Publish( MQTTContext_t * const pContext,
 
     if( status == MQTTSuccess )
     {
-        /* Send header first. */
-        bytesSent = sendPacket( pContext,
-                                pContext->networkBuffer.pBuffer,
-                                headerSize );
-
-        if( bytesSent < 0 )
+        /* Reserve state for publish message. Only to be done for QoS1 or QoS2. */
+        if( pPublishInfo->qos > MQTTQoS0 )
         {
-            LogError( "Transport send failed for PUBLISH header." );
-            status = MQTTSendFailed;
+            status = MQTT_ReserveState( pContext,
+                                        packetId,
+                                        pPublishInfo->qos );
         }
-        else
-        {
-            LogDebugWithArgs( "Sent %d bytes of PUBLISH header.",
-                              bytesSent );
+    }
+
+    if( status == MQTTSuccess )
+    {
+        /* Sends the serialized publish packet over network. */
+        status = sendPublish( pContext,
+                              pPublishInfo,
+                              headerSize );
+    }
 
-            /* Send Payload. */
-            bytesSent = sendPacket( pContext,
-                                    pPublishInfo->pPayload,
-                                    pPublishInfo->payloadLength );
-
-            if( bytesSent < 0 )
-            {
-                LogError( "Transport send failed for PUBLISH payload." );
-                status = MQTTSendFailed;
-            }
-            else
-            {
-                LogDebugWithArgs( "Sent %d bytes of PUBLISH payload.",
-                                  bytesSent );
-            }
+    if( status == MQTTSuccess )
+    {
+        /* Update state machine after PUBLISH is sent.
+         * Only to be done for QoS1 or QoS2. */
+        if( pPublishInfo->qos > MQTTQoS0 )
+        {
+            status = MQTT_UpdateStatePublish( pContext,
```

Comment:
`MQTT_UpdateStatePublish` returns an `MQTTPublishState_t`, not an `MQTTStatus_t`. This is done so that the library knows the next state of incoming publishes, and whether it should send a PUBACK or PUBREC. However, we might want to refactor it to return a status and have the publish state as an output parameter. Or, since these functions are internal, we can split them into separate functions for incoming and outgoing publishes (I think this option would be better since it will result in the simplest functions). If we want to do that, then it should be done before this PR is merged.

Reply:
I agree a refactor would be good. I think definitely the return type has to be addressed in this PR itself. Thanks for that catch.  This PR doesn't need to be blocked for that change.  Depending on which PR gets in first, It will just have to be handled in the rebase.